Implementation of multiple targets based on PR #404 and #396

dereulenspiegel opened the pull request:
This Pull Request implements multiple targets for endpoints based on #404 and #396 and therefore depends on #404. 
The following has been done so far:
- [x] Create type `endpoint.Targets` representing multiple targets (just like `endpoint.Labels`)
- [x] Modify every source and provider to use the new type (but without the extra logic to actually have multiple targets per endpoint).
- [x] Modify plan and conflict resolver to be handle to the new `endpoint.Targets`.
- [x] Modify the ingress source to actually generate endpoints with multiple targets.
- [x] Modify the Google Cloud DNS provider to make use of multiple targets for A records.
- [x] Modify the service source to generate endpoints with multiple targets.

This is what is still missing:
- [ ] Modify every provider except for the Google Cloud DNS provider to properly handle endpoints with multiple targets.
- [ ] Modify `ConflictResolver` and probably `Plan` to make more sensible decisions based on endpoints with multiple targets.

Currently this code runs in our GKE based cluster and works as expected.

ideahitme commented:
please rebase on master :)

dereulenspiegel commented:
Done :)

ideahitme commented on endpoint/endpoint.go:
```diff
@@ -38,12 +39,78 @@ func (ttl TTL) IsConfigured() bool {
+type Targets []string
+
+// NewTargets is a convenience method to create a new Targets object from a vararg of strings
+func NewTargets(target ...string) Targets {
```
Is the point here to create a copy list of strings into another slice ? otherwise you could also simply do 
```
func NewTargets(target ...string) Targets {
 	return target
}
```
Or even `Targets([]string{"1.1.1.1"})`

dereulenspiegel replied:
Yeah, the goal was to actually make copy. I read that there is still discussion about how to store DNS names (with or without trailing dot) internally. This can give the opportunity to modify the target to our internally used format, without modifying the original slice.
But I don't really know how useful this is or can be.

ideahitme replied:
At this point in the code, I think it is safe not to create a copy and use whatever is generated by sources (trailing dots are trimmed at that stage). It is DNS provider's responsibility to add or remove trailing dots when they are applied.

ideahitme commented on endpoint/endpoint.go:
```diff
@@ -38,12 +39,78 @@ func (ttl TTL) IsConfigured() bool {
+// as either being the shorter list of targets or where the first entry is less.
+// FIXME We really need to define under which circumstances a list Targets is considered 'less'
+// than another.
+func (t Targets) IsLess(o Targets) bool {
```
👍

szuecs replied:
Name feels a bit awkward, but I am fine with it for this PR if noone has a better name (I don't have right now)

linki replied:
I'm fine with the name. I'm wondering about the fixme: It says we should define what less means but at the same time describes it above. Is that FIXME still valid?

dereulenspiegel replied:
I think so. At least in the sense that current description is a description of what I implemented and nothing which was discussed. This related pretty directly to my comments on the ConflictResolver.
In the past `less` was something related to string comparison and was clearly defined by GoLang. Now we have something more complex and need to define when a list of targets is `less` than the other. In my opinion this is not as clear as string comparison.
It may be that the current definition is completely fine, then I will remove the FIXME, but I am very open to ideas for the definition if `less`.

ideahitme commented on provider/google.go:
```diff
@@ -175,13 +175,19 @@ func (p *GoogleProvider) Records() (endpoints []*endpoint.Endpoint, _ error) {
+			}
 			for _, rr := range r.Rrdatas {
 				// each page is processed sequentially, no need for a mutex here.
-				if supportedRecordType(r.Type) {
```
why was this check dropped ?

dereulenspiegel replied:
Because the check is now in the enclosing for loop. We need to check only once the type of the Resource Record and then I simply append all data in the Resource Record as targets.

ideahitme replied:
you are right, my bad :)

ideahitme commented:
looks like a good start, let me know when you need a review

dereulenspiegel commented:
Thanks. Let me update the Service source first, this is still missing. But how do we handle the situation with the providers? Currently only the google providers takes advantage of the new feature. The others providers should behave as before (no guarantees). Do we need to modify each and every provider before this feature can be merged (or even reviewed)?

ideahitme commented on provider/google.go:
```diff
@@ -322,9 +328,10 @@ func newRecord(ep *endpoint.Endpoint) *dns.ResourceRecordSet {
 	// way we can use it has is here and trim it off if it exists when necessary.
-	target := ep.Target
+	targets := make([]string, len(ep.Targets))
+	copy(targets, []string(ep.Targets))
```
I don't think there is a need to create a copy, after this stage the whole chunk of memory should be released and used no more.

ideahitme replied:
It is also worth mentioning in the comment, that the `Targets` list comes sorted

dereulenspiegel replied:
Interestingly removing the copy causes quite some failures in tests. The tests are now complaining that expected and actual Endpoints don't match. I have to take a closer look at this.

szuecs replied:
Is this checking a TODO? @dereulenspiegel

ideahitme commented:
@dereulenspiegel we don't need to support multi-target in all providers, however they should be working. 

One thing to note is that we should probably be sorting all targets per endpoint, to guarantee consistency in providers for which we pick first target

ideahitme commented:
@dereulenspiegel could you please answer my last comment:

> One thing to note is that we should probably be sorting all targets per endpoint, to guarantee consistency in providers for which we pick first target

dereulenspiegel commented:
@ideahitme Sorry for the late reply. I started the year off with a nice infection. 
I implemented sorting in the service and ingress sources as well as the Google Cloud DNS provider. This should make sure that we have a predictable order of Targets in all places where we actually use multiple targets. Also I rebased to the current master, therefore all the "new" commits.

ideahitme commented:
@dereulenspiegel awesome, I will take a look in the nearest future. @linki please take a look as well

szuecs commented:
@dereulenspiegel I guess now all the "still missing" checkboxes are fixed, right?

szuecs commented on endpoint/endpoint_test.go:
```diff
@@ -22,15 +22,15 @@ import (
-	if w.DNSName != "example.org" || w.Target != "load-balancer.com" || w.RecordType != "" {
+	if w.DNSName != "example.org" || w.Targets[0] != "load-balancer.com" || w.RecordType != "" {
 		t.Error("endpoint is not initialized correctly")
 	}
```
Can you also add additional tests that show the working new functionality, please?

szuecs commented on source/ingress.go:
```diff
@@ -109,33 +110,33 @@ func (sc *ingressSource) Endpoints() ([]*endpoint.Endpoint, error) {
 
+	for _, ep := range endpoints {
+		sort.Sort(ep.Targets)
+	}
```
Why this needs to be sorted?
Does anyone rely on having this slice sorted?

dereulenspiegel replied:
Currently probably not, but for the planning component it would make sense to have an ordered list of targets. It was just my guess that it is better to have every source return a sorted list oft targets as a convention early on. And on the other it probably won't hurt to have the targets sorted.

szuecs replied:
Ok, I have no real opinion on that.

szuecs commented on source/ingress.go:
```diff
@@ -109,33 +110,33 @@ func (sc *ingressSource) Endpoints() ([]*endpoint.Endpoint, error) {
-func getEndpointsFromTargetAnnotation(ing *v1beta1.Ingress, hostname string) []*endpoint.Endpoint {
-	var endpoints []*endpoint.Endpoint
+// FIXME(dereulenspiegel) For now we won't use this method.
+func getTargetsFromTargetAnnotation(ing *v1beta1.Ingress) endpoint.Targets {
```
FIXME: do you use this function? 
If not delete it and have it locally stashed if you do not want to loose it.

szuecs commented on source/ingress.go:
```diff
@@ -145,26 +146,18 @@ func (sc *ingressSource) endpointsFromTemplate(ing *v1beta1.Ingress) ([]*endpoin
-			endpoints = append(endpoints, endpoint.NewEndpointWithTTL(hostname, lb.Hostname, endpoint.RecordTypeCNAME, ttl))
-		}
+
+	targets := getTargetsFromTargetAnnotation(ing)
```
here you use the function as far as I see, so drop the fixme :)

dereulenspiegel replied:
Still, I removed the unnecessary comment :)

szuecs commented on source/ingress.go:
```diff
@@ -145,26 +146,18 @@ func (sc *ingressSource) endpointsFromTemplate(ing *v1beta1.Ingress) ([]*endpoin
+	targets := getTargetsFromTargetAnnotation(ing)
+
+	if len(targets) == 0 {
+		targets = targetsFromLb(ing.Status)
```
targetsFromLb() the naming is a bit weird, better: targetsFromIngressStatus()

szuecs commented on source/ingress.go:
```diff
@@ -208,32 +201,76 @@ func (sc *ingressSource) setResourceLabel(ingress v1beta1.Ingress, endpoints []*
+	var cnameTargets endpoint.Targets
+
+	for _, t := range targets {
+		if suitableType(t) == endpoint.RecordTypeA {
```
for type checking I prefer a switch case block and use default as else

szuecs commented on source/ingress.go:
```diff
@@ -208,32 +201,76 @@ func (sc *ingressSource) setResourceLabel(ingress v1beta1.Ingress, endpoints []*
 	return endpoints
 }
+
+func targetsFromLb(status v1beta1.IngressStatus) endpoint.Targets {
```
as already stated, better naming targetsFromLb -> targetsFromIngressStatus

szuecs commented on source/service.go:
```diff
@@ -119,10 +120,14 @@ func (sc *serviceSource) Endpoints() ([]*endpoint.Endpoint, error) {
 	}
 
+	for _, ep := range endpoints {
+		sort.Sort(ep.Targets)
```
does anyone rely on this being sorted?

szuecs commented on source/service.go:
```diff
@@ -134,6 +139,7 @@ func (sc *serviceSource) extractHeadlessEndpoint(svc *v1.Service, hostname strin
 	}
 
 	for _, v := range pods.Items {
+
```
empty line, please drop that :)

szuecs commented on source/service.go:
```diff
@@ -218,54 +224,85 @@ func (sc *serviceSource) setResourceLabel(service v1.Service, endpoints []*endpo
+		RecordTTL:  ttl,
+		RecordType: endpoint.RecordTypeA,
+		Labels:     endpoint.NewLabels(),
+		Targets:    make(endpoint.Targets, 0, 10),
```
10 -> please use a constant

szuecs commented on source/service.go:
```diff
@@ -218,54 +224,85 @@ func (sc *serviceSource) setResourceLabel(service v1.Service, endpoints []*endpo
+		RecordTTL:  ttl,
+		RecordType: endpoint.RecordTypeCNAME,
+		Labels:     endpoint.NewLabels(),
+		Targets:    make(endpoint.Targets, 0, 10),
```
use the same constant as above

szuecs commented on source/service.go:
```diff
@@ -218,54 +224,85 @@ func (sc *serviceSource) setResourceLabel(service v1.Service, endpoints []*endpo
-			endpoints = append(endpoints, sc.extractHeadlessEndpoint(svc, hostname)...)
+			endpoints = append(endpoints, sc.extractHeadlessEndpoints(svc, hostname)...)
+		}
+
```
drop empty line

linki commented:
Merged-in current master branch

dereulenspiegel commented:
@szuecs I should now have fixed the nitpicks (actually pretty important to enforce consistent coding style and quality) or have commented on the nitpicks if I had comments. Also I rebased to the current state of master.
Except for modifying existing provider (except for the Google Cloud DNS provider) and checking if the algorithm in `ConflictResolver` still makes sense everything should be done.
As for the algorithm in `ConflictResolver` I am not even sure if it needs to be changed, but since I am not really an expert in DNS best practices I added this as a ToDo just so someone more experienced would check.

szuecs commented:
@dereulenspiegel thanks I will check

szuecs commented:
lgtm, I have nothing to add 
ping @ideahitme @linki

linki commented on internal/testutils/endpoint.go:
```diff
@@ -33,21 +33,20 @@ func (b byAllFields) Less(i, j int) bool {
-		}
-		if b[i].Target == b[j].Target {
+		// This rather bad, we need a more complex comparison for Targets, which considers all elements
+		if b[i].Targets.Same(b[j].Targets) {
```
How bad is this "rather bad" part? Could it lead to tests not failing when they better should?

dereulenspiegel replied:
Probably not that bad. I had no failing tests so far, and afaik this comparison should be enough for testing. Should I removed the comment or do we want to keep it for the future until all providers have been modified for multiple targets in case we discover problems then?

linki replied:
Let's keep it.

linki commented on provider/aws.go:
```diff
@@ -391,7 +391,7 @@ func newChange(action string, endpoint *endpoint.Endpoint) *route53.Change {
 		change.ResourceRecordSet.ResourceRecords = []*route53.ResourceRecord{
 			{
-				Value: aws.String(endpoint.Target),
+				Value: aws.String(endpoint.Targets[0]),
```
To make multi-target really work this should actually create a `ResourceRecord` for each `Targets[i]` I believe. I agree to not have this in-scope for this PR. Just want to raise awareness.

linki commented on source/service.go:
```diff
@@ -221,54 +231,85 @@ func (sc *serviceSource) setResourceLabel(service v1.Service, endpoints []*endpo
-	if err != nil {
-		log.Warn(err)
-	}
+func extractServiceIps(svc *v1.Service) endpoint.Targets {
```
@dereulenspiegel Can you double-check this function. Seems to me it only returns an empty list or the cluster IP and can be simplified to:

```golang
if svc.Spec.ClusterIP != v1.ClusterIPNone {
  return endpoint.Targets{svc.Spec.ClusterIP}
}
return nil // or endpoint.Targets{}
```

dereulenspiegel replied:
Yeah, you're right. Sometimes I don't see the woods because of all of the trees. I fixed it.

linki replied:
Me too :sweat_smile: I changed it one more time to make it easier for my brain.

linki commented:
@dereulenspiegel Impressive, thank you! I left some things to double-check for your. Apart from that it lgtm.

dereulenspiegel commented:
@linki Thanks. I hope I checked and fixed now everything preventing a merge. If there is still something left I overlooked please notify me.

linki commented:
@dereulenspiegel Can you also add a changelog entry to the top of the `CHANGELOG` file. Please also state which resource type and provider support multiple targets with this change. Don't forget the PR number and your name.

dereulenspiegel commented:
@linki Done :)

linki commented:
@dereulenspiegel Awesome, thanks a lot. This is an importatn milestone! Thanks to all the reviewers as well, e.g. @szuecs @ideahitme 🎉 🎈

linki commented:
For anyone interested in trying this out it's available on our Docker registry:
```
docker run registry.opensource.zalan.do/teapot/external-dns:v0.5.0-alpha.1 --help
```

So far it behaves exactly as it should on our GKE cluster.
